### PR TITLE
bug: fix CSI driver inline implementation

### PR DESCRIPTION
**Pull request by helayoty**
Update the Azure file CSI driver inline volume support: https://docs.microsoft.com/en-us/azure/aks/azure-files-volume#mount-file-share-as-an-inline-volume

### CODE REVIEW

File: provider/aci.go [hunk 1]
```diff
@@ -1639,38 +1640,39 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
-		if err != nil {
-			return nil, errors.Wrap(err, fmt.Sprintf("The secret %s for AzureFile CSI driver %s is not found", volume.CSI.NodePublishSecretRef.Name, volume.Name))
-		}
+	azureSource := &aci.AzureFileVolume{}
 
-		if secret == nil {
-			return nil, fmt.Errorf("Getting secret for AzureFile CSI driver %s returned an empty secret", volume.Name)
-		}
+	if volume.CSI.VolumeAttributes != nil && len(volume.CSI.VolumeAttributes) != 0 {
+		secretName := volume.CSI.VolumeAttributes[azureFileSecretName]
```
Comment (andyzhangx) [inline provider/aci.go:1646]: field name in `CSI.VolumeAttributes` is case insensitive, you could refer to following code to handle such issue:

https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/dabcf18dfa32fac089f45c182b63726726bc29b1/pkg/azurefile/azurefile.go#L568-L588

File: provider/aci.go [hunk 2]
```diff
@@ -1639,38 +1640,39 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
-		// Set shareName
-		if volume.CSI.VolumeAttributes != nil {
-			if shareName, ok := volume.CSI.VolumeAttributes[azureFileShareName]; ok {
-				azureSource.ShareName = shareName
+			if secret == nil {
+				return nil, fmt.Errorf("getting secret for AzureFile CSI driver %s returned an empty secret", volume.Name)
 			}
+			// Set the storage account name and key
+
+			azureSource.StorageAccountName = string(secret.Data["azurestorageaccountname"])
```
Comment (andyzhangx) [inline provider/aci.go:1659]: use `strings.TrimSpace()`

File: provider/aci.go [hunk 3]
```diff
@@ -1639,38 +1640,39 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
+			azureSource.StorageAccountKey = string(secret.Data["azurestorageaccountkey"])
```
Comment (andyzhangx) [inline provider/aci.go:1660]: it's better define `azurestorageaccountkey` and `azurestorageaccountname` as const vars since they are referenced multiple times.

File: provider/aci.go [hunk 4]
```diff
@@ -1639,38 +1642,39 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
-		if err != nil {
-			return nil, errors.Wrap(err, fmt.Sprintf("The secret %s for AzureFile CSI driver %s is not found", volume.CSI.NodePublishSecretRef.Name, volume.Name))
-		}
+	azureSource := &aci.AzureFileVolume{}
 
-		if secret == nil {
-			return nil, fmt.Errorf("Getting secret for AzureFile CSI driver %s returned an empty secret", volume.Name)
-		}
+	if volume.CSI.VolumeAttributes != nil && len(volume.CSI.VolumeAttributes) != 0 {
+		secretName := volume.CSI.VolumeAttributes[azureFileSecretName]
```
Comment (andyzhangx) [inline provider/aci.go:1648]: it's case insensitive, not lower case, follow this code example to handle such case: https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/dabcf18dfa32fac089f45c182b63726726bc29b1/pkg/azurefile/azurefile.go#L568-L588

File: provider/aci.go [hunk 5]
```diff
@@ -1639,38 +1642,39 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
-		// Set shareName
-		if volume.CSI.VolumeAttributes != nil {
-			if shareName, ok := volume.CSI.VolumeAttributes[azureFileShareName]; ok {
-				azureSource.ShareName = shareName
+			if secret == nil {
+				return nil, fmt.Errorf("getting secret for AzureFile CSI driver %s returned an empty secret", volume.Name)
 			}
+			// Set the storage account name and key
+
+			azureSource.StorageAccountName = strings.TrimSpace(strings.ToLower(string(secret.Data[azureFileStorageAccountName])))
```
Comment (andyzhangx) [inline provider/aci.go:1661]: remove `strings.ToLower`

File: provider/aci.go (hunk 4 above)
Comment (helayoty) [reply provider/aci.go]: Got it, please let me know how you think now.

**Review by mingqishao — DISMISSED** [review]
look good for me.

File: provider/aci.go [hunk 6]
```diff
@@ -1639,38 +1642,39 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
-	}
-	if volume.CSI.NodePublishSecretRef != nil && volume.CSI.NodePublishSecretRef.Name != "" {
+	azureSource := &aci.AzureFileVolume{}
 
-		// Set the storage account name and key
-		secret, err := p.resourceManager.GetSecret(volume.CSI.NodePublishSecretRef.Name, namespace)
-		if err != nil {
-			return nil, errors.Wrap(err, fmt.Sprintf("The secret %s for AzureFile CSI driver %s is not found", volume.CSI.NodePublishSecretRef.Name, volume.Name))
-		}
+	if volume.CSI.VolumeAttributes != nil && len(volume.CSI.VolumeAttributes) != 0 {
```
Comment (andyzhangx) [inline provider/aci.go:1647]: I would suggest write like this, this would check field name first, also add a shareName filed check.
```
var secretName, shareName string
if volume.CSI.VolumeAttributes != nil {
	for k, v := range volume.CSI.VolumeAttributes {
		switch strings.ToLower(k) {
		case azureFileSecretName:
			secretName := v
		case azureFileShareName:
			shareName = v
	}
}

if shareName == "" {
	return error ...
}
azureSource.ShareName = shareName

if secretName == "" {
	return error ...
}

secret, err := p.resourceManager.GetSecret(secretName, namespace)
if err != nil {
	return nil, errors.Wrap(err, fmt.Sprintf("the secret %s for AzureFile CSI driver %s is not found", secretName, volume.Name))
}
if secret == nil {
	return nil, fmt.Errorf("getting secret for AzureFile CSI driver %s returned an empty secret", volume.Name)
}

// Set the storage account name and key
azureSource.StorageAccountName = strings.TrimSpace(string(secret.Data[azureFileStorageAccountName]))
azureSource.StorageAccountKey = strings.TrimSpace(string(secret.Data[azureFileStorageAccountKey]))
```

**Review by andyzhangx — DISMISSED** [review]
/lgtm

File: provider/aci.go [hunk 7]
```diff
@@ -1639,38 +1642,45 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("the secret %s for AzureFile CSI driver %s is not found", secretName, volume.Name))
 	}
+
+	if secret == nil {
+		return nil, fmt.Errorf("getting secret for AzureFile CSI driver %s returned an empty secret", volume.Name)
+	}
+
+	// Set the storage account name and key
+	azureSource.StorageAccountName = strings.TrimSpace(string(secret.Data[azureFileStorageAccountName]))
```
Comment (ryanzhang-oss) [inline provider/aci.go:1678]: what if the secret doesn't contain that azureFileStorageAccountName key?

Comment (helayoty) [reply provider/aci.go:1682]: Good call. Added this corner case along with the unit tests

File: provider/aci.go [hunk 8]
```diff
@@ -1639,38 +1642,49 @@ func getProbe(probe *v1.Probe, ports []v1.ContainerPort) (*aci.ContainerProbe, e
...
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("the secret %s for AzureFile CSI driver %s is not found", secretName, volume.Name))
+	}
+
+	if secret == nil {
+		return nil, fmt.Errorf("getting secret for AzureFile CSI driver %s returned an empty secret", volume.Name)
 	}
+
+	// Set the storage account name and key
+	if secret.Data[azureFileStorageAccountName] == nil || secret.Data[azureFileStorageAccountKey] == nil {
```
Comment (andyzhangx) [inline provider/aci.go:1678]: the `secret.Data[azureFileStorageAccountName]` actually returns `[]byte`, and `string([]byte{})` return empty string, so actually that is not a problem, no panic if no check.

`Data map[string][]byte`

Comment (helayoty) [reply provider/aci.go:1678]: @andyzhangx So no change is required?

Comment (andyzhangx) [reply provider/aci.go:1678]: yes, that check is not necessary, let's make code clean

Comment (helayoty) [reply provider/aci.go]: done

File: provider/aci.go (hunk 7 above)
Comment (helayoty) [reply provider/aci.go:1678]: @ryanzhang-oss  according to Andy's comment [here](https://github.com/virtual-kubelet/azure-aci/pull/199#discussion_r870896485), no need for this check.

**Review by ryanzhang-oss — APPROVED** [review]
